README.md

# JQuantsR

<!-- badges: start -->
<!-- badges: end -->

[J-Quants
API](https://application.jpx-jquants.com/)は、日本取引所グループ（JPX）が[J-Quants](https://jpx-jquants.com/)の取り組みの一環として提供する株式情報APIです。

JQuantsRは、J-Quants
APIをRから扱いやすいようにラップしたRパッケージです。

## インストール

以下の通り、本リポジトリからインストールしてください。

インストールには4.0.0以上のバージョンのRが必要です。

``` r
install.packages("remotes")
remotes::install_github("J-Quants/JQuantsR")
```

## ユーザー登録

JQuantsRを使用する前に、[J-Quants
API](https://application.jpx-jquants.com/)よりユーザー登録を行ってください。

ユーザー登録後、ログインしてリフレッシュトークンを取得してください。

## サンプルコード

``` r
# 各Rセッションの最初にJQuantsR::authorize()を実行してください。
# Rの環境変数"JQUANTSR_REFRESH_TOKEN"にリフレッシュトークンを設定すると、
# JQuantsR::authorize(refresh_token = "YOUR REFRESH TOKEN")の代わりに
# JQuantsR::authorize()で実行できます。
JQuantsR::authorize(refresh_token = "YOUR REFRESH TOKEN")

# 引数や返り値の詳細はJ-Quants API Referenceや関数のヘルプをご参照ください。
JQuantsR::get_info()
JQuantsR::get_info(code = "86970")
JQuantsR::get_sections()
JQuantsR::get_daily_quotes(code = "86970")
JQuantsR::get_daily_quotes(code = "86970", from = "20220701", to = "20220715")
JQuantsR::get_daily_quotes(date = "20220701")
JQuantsR::get_financial_statements(code = "86970")
JQuantsR::get_financial_announcement()
```

## 各関数について

### 概要

本パッケージに含まれる関数は以下の通りです。

-   `JQuantsR::authorize()`:
    リフレッシュトークンを用いてIDトークンを取得する
    -   POST to “<https://api.jpx-jquants.com/v1/token/auth_refresh>”
-   `JQuantsR::get_info()`: 最新の全上場銘柄一覧を取得する
    -   GET to “<https://api.jpx-jquants.com/v1/listed/info>”
-   `JQuantsR::get_sections()`: 業種一覧を取得する
    -   GET to “<https://api.jpx-jquants.com/v1/listed/sections>”
-   `JQuantsR::get_daily_quotes()`: 日次の株価を取得する
    -   GET to “<https://api.jpx-jquants.com/v1/prices/daily_quotes>”
-   `JQuantsR::get_financial_statements()`: 決算情報を取得する
    -   GET to “<https://api.jpx-jquants.com/v1/fins/statements>”
-   `JQuantsR::get_financial_announcement()`:
    翌日の決算発表予定銘柄を取得する
    -   GET to “<https://api.jpx-jquants.com/v1/fins/annoucement>”

`JQuantsR::authorize()`を除き、取得したデータをtibbleで返します。

各関数の引数や返り値、取得できるデータの詳細は、[J-Quants API
Reference](https://jpx.gitbook.io/j-quants-api/api-reference)やJQuantsRのヘルプをご参照ください。また、データの更新時間は[データの更新頻度](https://jpx.gitbook.io/j-quants-api/api-reference/data-update)をご参照ください。

### `JQuantsR::authorize()`

-   各Rセッションにおいて、最初に`JQuantsR::authorize()`を実行してIDトークンを取得する必要があります。
    -   `JQuantsR::authorize()`は、IDトークンを取得して`JQUANTSR_ID_TOKEN`という変数名の環境変数にIDトークンをセットします。
    -   各Rセッション内で一度`JQuantsR::authorize()`を実行すれば、`id_token`を引数に取る各関数にIDトークンを渡す必要はありません。
-   リフレッシュトークンを`JQUANTSR_REFRESH_TOKEN`という変数名の環境変数に設定すれば、`JQuantsR::authorize()`の引数`refresh_token`にリフレッシュトークンを指定する必要はありません。
    -   `.Renviron`ファイルに`JQUANTSR_REFRESH_TOKEN`を記載することを推奨します。
-   リフレッシュトークンとIDトークンの有効期限はそれぞれ1週間、24時間です。
    -   リフレッシュトークンが有効期限切れの場合は、[J-Quants
        API](https://application.jpx-jquants.com/)にログインしてリフレッシュトークンを取得してください。
    -   IDトークンが有効期限切れの場合は、`JQuantsR::authorize()`を実行してIDトークンを取得してください。

## その他

-   J-Quants
    APIは現在ベータ版であり、また本パッケージは現在開発途上であるため、本パッケージの仕様は今後変更される可能性があります。
-   バグ報告や機能追加のリクエストなどはIssuesにご投稿ください。Pull
    Requestも歓迎です。

## Reference

-   [J-Quants](https://jpx-jquants.com/)
-   [J-Quants API](https://application.jpx-jquants.com/)
-   [J-Quants API
    Reference](https://jpx.gitbook.io/j-quants-api/api-reference)
    -   [データの更新頻度](https://jpx.gitbook.io/j-quants-api/api-reference/data-update)


NEWS.md
# JQuantsR 0.1.0

- JQuantsRを公開しました。


R/get_api.R
#' Get data from J-Quants API
#'
#' @param code a string stock code.
#' @param from a string the start date of data. "\%Y\%m\%d" and "\%Y-\%m-\%d" formats are valid.
#' @param to a string the end date of data. "\%Y\%m\%d" and "\%Y-\%m-\%d" formats are valid.
#' @param date a string the date of data. "\%Y\%m\%d" and "\%Y-\%m-\%d" formats are valid.
#' @param id_token a string your id_token.
#' @return if successfully fetched data (= status code of the API is 200), a tibble of the fetched data.
#' if not successfully (= status code of the API is not 200), none and message from the API is printed.
#' @details parameter "code", "from", "to", "date" are required as follows.
#' \itemize{
#'   \item{"get_info()"}: {"code" is optional.}
#'   \item{"get_daily_quotes()"}: {At least one of "code" and "date" are mandatory. If "from" and "to" parameters are used, "code" is mandatory.}
#'   \item{"get_financial_annoucements()"}: {At least one of "code" and "date" are mandatory.}
#' }
#' For more information, see \href{https://jpx.gitbook.io/j-quants-api/api-reference}{J-Quants API reference}.
#'
#' parameter "id_token": When you have run "authorize()", you don't need to set "id_token".
#'
#' @examples
#' \dontrun{
#' # you must have done "authorize()" before running "get_info" or other functions,
#  # if you don't pass "id_token" in such functions.
#' authorize(refresh_token = YOUR_REFRESH_TOKEN)
#' get_info()
#' get_info(code = "86970")
#' get_sections()
#' get_daily_quotes(date = "20220701")
#' get_daily_quotes(code = "86970", from = "20220101", to = "20220630")
#' get_financial_statements(code = "86970")
#' get_financial_statements(date = "20220105")
#' get_financial_annoucement()
#' }
#' @export
get_info <- function(code, id_token = Sys.getenv("JQUANTSR_ID_TOKEN")) {
  if (rlang::is_missing(code)) {
    code <- NULL
  }
  query <- list(code = code)
  get_from_api("/listed/info", query, id_token)
}

#' @rdname get_info
#' @export
get_sections <- function(id_token = Sys.getenv("JQUANTSR_ID_TOKEN")) {
  query <- NULL
  get_from_api("/listed/sections", query, id_token)
}

#' @rdname get_info
#' @export
get_daily_quotes <- function(code, from, to, date, id_token = Sys.getenv("JQUANTSR_ID_TOKEN")) {
  if (rlang::is_missing(code)) {
    code <- NULL
  }
  if (rlang::is_missing(from)) {
    from <- NULL
  }
  if (rlang::is_missing(to)) {
    to <- NULL
  }
  if (rlang::is_missing(date)) {
    date <- NULL
  }
  query <- list(code = code, from = from, to = to, date = date)
  get_from_api("/prices/daily_quotes", query, id_token)
}

#' @rdname get_info
#' @export
get_financial_statements <- function(code, date, id_token = Sys.getenv("JQUANTSR_ID_TOKEN")) {
  if (rlang::is_missing(code)) {
    code <- NULL
  }
  if (rlang::is_missing(date)) {
    date <- NULL
  }
  query <- list(code = code, date = date)
  get_from_api("/fins/statements", query, id_token)
}

#' @rdname get_info
#' @export
get_financial_announcement <- function(id_token = Sys.getenv("JQUANTSR_ID_TOKEN")) {
  query <- NULL
  get_from_api("/fins/announcement", query, id_token)
}
